src/node_id.rs

```diff
@@ -0,0 +1,334 @@
+    pub fn to_base58(&self) -> String {
+        bs58::encode(self.discohash.as_slice()).into_string()
+    }
+
+    #[inline]
+    fn digest(&self) -> &[u8] {
+        &self.discohash.as_slice()
+    }
+
+    // TODO: better verification that `data` is a valid `NodeId`
+    // would require logic from `is_public_key`
+    #[inline]
+    fn from_bytes(data: Vec<u8>) -> Result<NodeId, ParseError> {
+        if data.len() != 32 {
+            return Err(ParseError);
+        }
+        let new_node = NodeId { discohash: data };
+        Ok(new_node)
+    }
+
+    #[inline]
+    fn is_public_key(&self, pubkey: PublicKey) -> bool {
+        let ret_val = true;
+        let mut counter = 0;
+        let pk_hash = hash(pubkey.as_bytes(), 32);
+        &pk_hash == &self.discohash
+    }
+
+    /// For Testing Purposes Only
+    ///
+    /// Note: does not generate keypair, just generates a random byte array
+    #[inline]
+    pub fn random(output_len: usize) -> NodeId {
+        NodeId {
+            discohash: DiscoHash::random(output_len),
+        }
+    }
+
+    /// Useful for distance metric and node_id generation checks
+    pub fn is_zero(&self) -> bool {
+        self.discohash.iter().all(|d| *d == 0)
+    }
+
+    /// Verify Signature
+    ///
+    /// verify message signature made with the public key associated with this NodeId
+    pub fn verify_msg(message: &[u8], signature: &[u8], public_key: &[u8]) -> bool {
+        // change receive type to an error on verifying signatures
+
+        // would need to use the sign method in Node
+        // or take a signature
+        todo!();
+    }
+}
+
+impl FromStr for NodeId {
+    type Err = ParseError;
+
+    #[inline]
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let bytes = bs58::decode(s).into_vec()?;
+        NodeId::from_bytes(bytes).map_err(|_| ParseError)
+    }
+}
+
+impl Ord for NodeId {
+    fn cmp(&self, other: &Self) -> Ordering {
+        self.discohash.cmp(&other.discohash)
+    }
+}
+
+impl PartialOrd for NodeId {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl From<PublicKey> for NodeId {
+    #[inline]
+    fn from(pubkey: PublicKey) -> NodeId {
+        let keyhash = hash(pubkey.as_bytes(), 32);
+        NodeId { discohash: keyhash }
+    }
+}
+
+impl PartialEq for NodeId {
+    fn eq(&self, other: &NodeId) -> bool {
+        self.discohash == other.discohash
+    }
+}
+
+impl PartialEq<NodeInfo> for NodeId {
+    fn eq(&self, other: &NodeInfo) -> bool {
+        *self == other.id
+    }
+}
+
+impl PartialEq<Vec<u8>> for NodeId {
+    fn eq(&self, other: &Vec<u8>) -> bool {
+        self.discohash == *other
+    }
+}
+
+pub trait KadMetric: PartialEq + Clone + fmt::Debug {
+    type Err;
+    type Metric;
+
+    fn distance(&self, other: &Self) -> Result<Self, Self::Err>;
+    // used in `store` specifically 
+    fn metric_distance(&self, other: &Self) -> Result<Self::Metric, Self::Err>;
+}
```

Comment:
How about renaming metric_distance to distance and impl From<Metric> for NodeId. You probably need that config trait we talked about:
```rust
pub trait Dht {
    type NodeId: From<Self::Metric>; 
    type Metric;
}
```

Reply:
Will do tomorrow -- wait for this before merging...

Reply:
I'm actually confused about what you're looking for here? Should I include the `distance` method in this new trait as well?

```rust
pub trait Dht {
    type Err;
    type NodeId: From<Self::Metric>;
    type Metric;

    fn distance(&self, other: &Self) -> Result<Self::Metric, Self::Err>;
}
```

And then, I guess I would implement this for `config::NodeHandler`, but that doesn't really feel right. It seems you have something in mind and I can't guess what it is.